### PR TITLE
Header Proof Validation

### PR DESCRIPTION
**Pull request by ScottyPoi**
Implements requesting and sharing of `header_proof` as a way to validate a `header_accumulator_snapshot`

*  Defines a `header_proof_interface`
   * epoch_hash
   * gindex
   * leaf
   * witnesses
* Defines a new SSZ container `header_proof`
  * Container [leaf: bytes32, witnesses: List [ bytes32, limit: 256  ] }
  * `leaf` is the hash_tree_root of the `HeaderRecord`
  * `witnesses` is the list of hashes for the merkle proof
 * Defines a new`HistoryNetworkContentType`
   * `HeaderProof = 5`
 * Creates method to generate proofs by blockhash
 * Creates method to validate proof against epoch_accumulator
 * Creates helper functions
   * `gIndexFromBlocknumber`
   * `epochIndexFromBlocknumber`
   * `listIndexFromBlocknumber`
 * Creates new method to validate snapshots
 * Defines a dynamic `threshold` for snapshot validation
 * Adds to `history_protocol` a record with one blockhash from each epoch by which to validate snapshots.
 * Implements validation scheme in snapshot reception method
 * Implements sending and receiving of `header_proof` requests through `FINDCONTENT/FOUNDCONTENT`

TODO: 
1. Decide what to do with un-verified snapshots.  
2. Decide the fate of the `header_proofs` after they serve their initial purpose
  - delete?
  - gossip?
  - store?
 3. Get the other `portal_network` teams feedback on the design

### CODE REVIEW

**Review by acolytec3 — COMMENTED** [review]
This is looking great so far!  Want me to jump in with tests or are you closing in on a finish?

File: packages/portalnetwork/src/subprotocols/protocol.ts [hunk 1]
```diff
@@ -31,6 +31,12 @@ import { randUint16, MAX_PACKET_SIZE } from '../wire/utp/index.js'
 import { RequestCode } from '../wire/utp/PortalNetworkUtp/PortalNetworkUTP.js'
 import { NodeLookup } from './nodeLookup.js'
 import { StateNetworkRoutingTable } from './state'
+import {
```
[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/portalnetwork/src/subprotocols/protocol.ts:34]: I was thinking the same thing.  `baseProtocol` should just be the abstract scaffold.

**Comment by ScottyPoi** [comment]
> This is looking great so far! Want me to jump in with tests or are you closing in on a finish?

No, I think I'm still honing in on the final form of the validation protocol in general.  

The pieces are in place, but we should discuss exactly what we're doing with the header proofs.

Currently we can do findContent for a `HeaderProof`, which is an ssz serialized container.  This isn't based on anything in the specs, it's adapted from the chainsafe ssz `Proof` interface.  We treat all proofs as `multiproofs`, and replace the `type` field with `epochRooot`.  The `epochRoot` field enables validation of `HeaderRecords` without any knowledge of the blockNumber, position in the epoch, or even the specific epoch index.   Also, the multiproof should enable us to send up to an entire epoch's (***8192***)   proofs in one `FoundContent` message, and offer up to ***26*** epoch's worth (for a grand total of...***212992*** header proofs!) 
```
{ epochroot,
gindices,
leaves,
witnesses,
}
```

**Comment by ScottyPoi** [comment]
I added a very unsophisticated protocol to `receiveSnapshot`, where before we take on a new snapshot as valid, we reach out (to the *network*) for a random set of `EpochAccumulators`, and a random set of `HeaderProofs`.  

Right now, nothing is in place to reject a snapshot that we can't validate, but the logger lets you know :stuck_out_tongue_winking_eye:

**Comment by ScottyPoi** [comment]
Since the `HeaderRecords` themselves are contained in the `HeaderProof` object, a node building an accumulator can gossip header proofs along with EpochAccumulators as the `HeaderAccumulator` is built.  Though as I write this, it feels like sending out a whole set of proofs is essentially gossiping bloated epoch accumulators.  

And requesting proofs for header records from an epoch accumulator we received doesn't really give us new information.

Seems to me that a snapshot could be validated simply by confirming a set of epoch accumulators, and if we really want, a set of header proofs for current epoch header records.

Maybe this only really adds anything if we do restrict it to "single proofs" after all.

**Comment by ScottyPoi** [comment]
I assume that we aren't actually *storing* these proofs in the database?  That they can be generated and gossiped effectively without.

**Comment by acolytec3** [comment]
> > This is looking great so far! Want me to jump in with tests or are you closing in on a finish?
> 
> No, I think I'm still honing in on the final form of the validation protocol in general.
> 
> The pieces are in place, but we should discuss exactly what we're doing with the header proofs.
> 
> Currently we can do findContent for a `HeaderProof`, which is an ssz serialized container. This isn't based on anything in the specs, it's adapted from the chainsafe ssz `Proof` interface. We treat all proofs as `multiproofs`, and replace the `type` field with `epochRooot`. The `epochRoot` field enables validation of `HeaderRecords` without any knowledge of the blockNumber, position in the epoch, or even the specific epoch index. Also, the multiproof should enable us to send up to an entire epoch's (_**8192**_) proofs in one `FoundContent` message, and offer up to _**26**_ epoch's worth (for a grand total of..._**212992**_ header proofs!)
> 
> ```
> { epochroot,
> gindices,
> leaves,
> witnesses,
> }
> ```

Isn't a multiproof with the entire epoch's worth of headers just the merkelized epoch accumulator?

**Comment by acolytec3** [comment]
> Since the `HeaderRecords` themselves are contained in the `HeaderProof` object, a node building an accumulator can gossip header proofs along with EpochAccumulators as the `HeaderAccumulator` is built. Though as I write this, it feels like sending out a whole set of proofs is essentially gossiping bloated epoch accumulators.

We shouldn't need both.  To be honest, you should never gossip a less than full epoch accumulator.  There isn't any real value in it that I can think of.  If someone wants the half-full accumulator, they can just request the individual headers and build it themselves.

> And requesting proofs for header records from an epoch accumulator we received doesn't really give us new information.
> 

Correct.

> Seems to me that a snapshot could be validated simply by confirming a set of epoch accumulators, and if we really want, a set of header proofs for current epoch header records.

Here again, you shouldn't need header proofs for the current epoch.  If you have the current epoch, you already know which headers are in it and can validate the contents of any individual header just by hashing it.

> Maybe this only really adds anything if we do restrict it to "single proofs" after all, and try to validate snapshots simply with these individual proofs, restricted to the set of blockhashes we already know. So a node receiving a snapshot could request proofs for headers that it already has

That would certainly be fruitful if you're trying to backfill historical epochs and fill out a header accumulator since you just use the root hash from each proof as that epoch's root.  Not sure what else it gains you though

**Comment by acolytec3** [comment]
> I assume that we aren't actually _storing_ these proofs in the database? That they can be generated and gossiped effectively without.

For now, no gain to us.

**Comment by ScottyPoi** [comment]
Cool thanks for the feedback.  Should be straightforward to wrap this up.
Should be able to simplify a bit if there's no obvious use for multiproofs.

Any opinion on what the threshold should be for trusting a snapshot?

**Comment by acolytec3** [comment]
Maybe if you can retrieve say three header proofs or epoch accumulators with correct root hashes (i.e. matches one in the snapshot)

File: packages/portalnetwork/src/subprotocols/protocol.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ScottyPoi) [reply packages/portalnetwork/src/subprotocols/protocol.ts:34]: I want to do this in a separate PR, though

**Comment by ScottyPoi** [comment]
Ok I think this is ready for another review @acolytec3

File: packages/portalnetwork/src/subprotocols/protocol.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (acolytec3) [reply packages/portalnetwork/src/subprotocols/protocol.ts:34]: ok, sounds good

File: packages/portalnetwork/src/subprotocols/history/history.ts [hunk 2]
```diff
@@ -152,7 +160,40 @@ export class HistoryProtocol extends BaseProtocol {
                     getHistoryNetworkContentId(1, 4),
                     decoded.value as Uint8Array
                   )
+                  break
                 }
+                // case HistoryNetworkContentTypes.HeaderProof: {
```
Comment (acolytec3) [inline packages/portalnetwork/src/subprotocols/history/history.ts:165]: Do we need any of this any more or can it be removed?

Comment (ScottyPoi) [reply packages/portalnetwork/src/subprotocols/history/history.ts]: Nope -- deleted

**Review by acolytec3 — APPROVED** [review]
Looks Awesome To Me.

**Pull request merged**